### PR TITLE
Implement PDF readers

### LINKED ISSUE #35: Implement PDFReader using Tabula
Use [tabula](https://github.com/tabulapdf/tabula-java) to extract data in tables from pdf and output to a model object.

### LINKED ISSUE #36: Implement PDFReader using Itext
Use [iText](https://github.com/itext/itext7) to extract text data from pdf and output to a model object.

### PR DESCRIPTION
**Pull request by Ts1n**
- Add [tabula](https://github.com/tabulapdf/tabula-java) and [Itext](https://github.com/itext/itext7) dependencies
- Add `PdfExtractor` interface and `PdfExtractorException`
- Add two `PdfExtractor` implementations, `TabulaPdfExtractor` and `ITextPdfExtractor`
- Add respective tests

Closes #35 , closes #36

### CODE REVIEW

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import com.itextpdf.kernel.pdf.PdfReader
+import com.itextpdf.kernel.pdf.canvas.parser.PdfTextExtractor
+import org.ionproject.integration.extractor.`interface`.PdfExtractor
+import org.ionproject.integration.extractor.exception.PdfExtractorException
+import org.ionproject.integration.utils.Try
+
+class ITextPdfExtractor : PdfExtractor {
+    /**
+     * Extract text data from pdf file locate at [pdfPath]
+     * @return Pair<Boolean, String>
```
Comment (sscosta) [inline src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt]: this might have been left from a previous version

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        for (i in 1..pdfDoc.numberOfPages) {
+            val result = Try.of { PdfTextExtractor.getTextFromPage(pdfDoc.getPage(i)) }
+            if (result is Try.Error) return Try.ofError(PdfExtractorException("Itext cannot process file"))
+
+            data.add((result as Try<String>).toString())
+        }
+
+        reader.close()
+
+        return Try.of(mutableListOf(data.toString()))
```
Comment (sscosta) [inline src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt]: as data is MutableList<String>, isn't this the same as "return Try.of(data)"?

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/TabulaPdfExtractor.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.commons.cli.DefaultParser
+import org.ionproject.integration.extractor.`interface`.PdfExtractor
+import org.ionproject.integration.extractor.exception.PdfExtractorException
+import org.ionproject.integration.utils.Try
+import technology.tabula.CommandLineApp
+
+class TabulaPdfExtractor : PdfExtractor {
+    /**
+     * Extract table data from pdf file locate at [pdfPath]
+     * @return Pair<Boolean, String>
```
Comment (sscosta) [inline src/main/kotlin/org/ionproject/integration/extractor/implementation/TabulaPdfExtractor.kt]: same as iText extractor

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/TabulaPdfExtractor.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        // Arguments to pass to tabula
+        // -g = Guess the portion of the page to analyze per page
+        // -l = Force PDF to be extracted using lattice-mode extraction
+        // -p = Page range
+        // -f = Output format JSON
+        val args = arrayOf(pdfFile.absolutePath, "-g", "-l", "-p", "all", "-f", "JSON")
+        val parser = DefaultParser()
+        val cmd = parser.parse(CommandLineApp.buildOptions(), args)
+
+        // CommandLineApp(data, cmd).extractTables(cmd)
```
Comment (sscosta) [inline src/main/kotlin/org/ionproject/integration/extractor/implementation/TabulaPdfExtractor.kt]: this might have been used in debug

**Review by sscosta — DISMISSED** [review]
saw some minor adjustments

**Review by sscosta — DISMISSED** [review]
Looks great!

File: src/main/kotlin/org/ionproject/integration/extractor/interface/PdfExtractor.kt [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+import org.ionproject.integration.utils.Try
+
+interface PdfExtractor {
+    /**
+     * Extract data from pdf file locate at [pdfPath]
+     * @return Try<MutableList<String>>
+     *     Success - String list contains extracted data
+     *     Failure - PdfExtractorException
+     */
+    suspend fun extract(pdfPath: String): Try<MutableList<String>>
```
Comment (pmhsfelix) [inline src/main/kotlin/org/ionproject/integration/extractor/interface/PdfExtractor.kt]: Why a mutable list?

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+     * Extract text data from pdf file locate at [pdfPath]
+     * @return Try<MutableList<String>>
+     *     Success - String list contains all extracted data in text format
+     *     Failure - PdfExtractorException
+     */
+    override suspend fun extract(pdfPath: String): Try<MutableList<String>> {
+        if (pdfPath.isEmpty()) return Try.ofError(PdfExtractorException("Empty path"))
+
+        val tryReader = Try.of { PdfReader(pdfPath) }
+        if (tryReader is Try.Error) return Try.ofError(PdfExtractorException("File doesn't exist or Invalid file format"))
```
Comment (pmhsfelix) [inline src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt]: We should need to do these explicit `ifs` when using `Try`. Did you face any blocking issues?

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        for (i in 1..pdfDoc.numberOfPages) {
+            val result = withContext(Dispatchers.IO) {
+                Try.of { PdfTextExtractor.getTextFromPage(pdfDoc.getPage(i)) }
+            }
+            if (result is Try.Error) return Try.ofError(PdfExtractorException("Itext cannot process file"))
+
+            data.add((result as Try<String>).toString())
+        }
+
+        reader.close()
```
Comment (pmhsfelix) [inline src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt]: Don't we need to ensure the `reader` is always closed?

File: src/main/kotlin/org/ionproject/integration/extractor/interface/PdfExtractor.kt (hunk 5 above)
Comment (pmhsfelix) [reply src/main/kotlin/org/ionproject/integration/extractor/interface/PdfExtractor.kt]: What is this  `pdfPath`? Can't it be a more strongly-typed _thing_ (e.g `URI` or `File`)?

File: src/test/kotlin/org/ionproject/integration/extractor/ITextPdfExtractorTests.kt [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        val pdfPath = "src/test/resources/test.pdf"
+        val result: Try<MutableList<String>>
+        val iTextPdfExtractor = ITextPdfExtractor()
+
+        // Act
+        result = runBlocking { iTextPdfExtractor.extract(pdfPath) }
+
+        // Assert
+        Assertions.assertEquals(true, result is Try.Value<MutableList<String>>)
+        Assertions.assertNotEquals(emptyList<String>(), (result as Try.Value<MutableList<String>>).value)
```
Comment (pmhsfelix) [inline src/test/kotlin/org/ionproject/integration/extractor/ITextPdfExtractorTests.kt]: We need a better way to do this

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import org.ionproject.integration.extractor.`interface`.PdfExtractor
+import org.ionproject.integration.extractor.exception.PdfExtractorException
+import org.ionproject.integration.utils.Try
+
+class ITextPdfExtractor : PdfExtractor {
+    /**
+     * Extract text data from pdf file locate at [pdfPath]
+     * @return Try<MutableList<String>>
+     *     Success - String list contains all extracted data in text format
+     *     Failure - PdfExtractorException
```
Comment (pmhsfelix) [inline src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt]: We shouldn't need to add these as comments. It should be explicit on the types. Perhaps we need to encode the exception type on the `Try`.

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/TabulaPdfExtractor.kt [hunk 10]
```diff
@@ -0,0 +1,44 @@
+package org.ionproject.integration.extractor.implementation
+
+import java.io.File
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.apache.commons.cli.DefaultParser
+import org.ionproject.integration.extractor.`interface`.PdfExtractor
```
Comment (pmhsfelix) [inline src/main/kotlin/org/ionproject/integration/extractor/implementation/TabulaPdfExtractor.kt]: Use plural (`interfaces`) to avoid using a reserved keyword.

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+        val tryReader = Try.of { PdfReader(pdfPath) }
+        if (tryReader is Try.Error) return Try.ofError(PdfExtractorException("File doesn't exist or Invalid file format"))
+
+        val reader = (tryReader as Try.Value<PdfReader>).value
+        val pdfDoc = PdfDocument(reader)
+
+        val data = mutableListOf<String>()
+        for (i in 1..pdfDoc.numberOfPages) {
+            val result = withContext(Dispatchers.IO) {
```
Comment (pmhsfelix) [inline src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt]: Why is this needed?
What are we trying to achieve with this `withContext`, since `PdfTextExtractor.getTextFromPage` is not suspend?

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/TabulaPdfExtractor.kt [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        // -g = Guess the portion of the page to analyze per page
+        // -l = Force PDF to be extracted using lattice-mode extraction
+        // -p = Page range
+        // -f = Output format JSON
+        val args = arrayOf(pdfFile.absolutePath, "-g", "-l", "-p", "all", "-f", "JSON")
+        val parser = DefaultParser()
+        val cmd = parser.parse(CommandLineApp.buildOptions(), args)
+
+        val data = StringBuilder()
+        var result = withContext(Dispatchers.IO) {
```
Comment (pmhsfelix) [inline src/main/kotlin/org/ionproject/integration/extractor/implementation/TabulaPdfExtractor.kt]: Same

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+import com.itextpdf.kernel.pdf.PdfDocument
+import com.itextpdf.kernel.pdf.PdfReader
+import com.itextpdf.kernel.pdf.canvas.parser.PdfTextExtractor
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.ionproject.integration.extractor.`interface`.PdfExtractor
+import org.ionproject.integration.extractor.exception.PdfExtractorException
+import org.ionproject.integration.utils.Try
+
+class ITextPdfExtractor : PdfExtractor {
```
Comment (pmhsfelix) [inline src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt]: Why does it start with an `I`?

File: src/main/kotlin/org/ionproject/integration/extractor/interface/PdfExtractor.kt (hunk 5 above)
Comment (Ts1n) [reply src/main/kotlin/org/ionproject/integration/extractor/interface/PdfExtractor.kt]: > 
> 
> Why a mutable list?

Data obtained via `iText` comes in string format and it's easier to process the data page by page.

> 
> 
> What is this `pdfPath`? Can't it be a more strongly-typed _thing_ (e.g `URI` or `File`)?

`pdfPath` is the path to the pdf file. It has to be a path because `Tabula` is executed as CLI.

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt (hunk 7 above)
Comment (Ts1n) [reply src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt:38]: We do. I'll fix it

File: src/test/kotlin/org/ionproject/integration/extractor/ITextPdfExtractorTests.kt (hunk 8 above)
Comment (Ts1n) [reply src/test/kotlin/org/ionproject/integration/extractor/ITextPdfExtractorTests.kt]: In what way?

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt (hunk 6 above)
Comment (Ts1n) [reply src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt]: `PdfReader(pdfPath)` throws an `java.io.IOException`, but we want all errors to be of type `PdfExtractorException` for easier error handling.

File: src/test/kotlin/org/ionproject/integration/extractor/ITextPdfExtractorTests.kt (hunk 8 above)
Comment (pmhsfelix) [reply src/test/kotlin/org/ionproject/integration/extractor/ITextPdfExtractorTests.kt]: Avoid the explicit cast, eventually by creating a helper function with it.

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt (hunk 9 above)
Comment (pmhsfelix) [reply src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt]: The returned type should document what are the returned options. Unfortunately, the `Try` design allows that for the success case but not for the error case.

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt (hunk 6 above)
Comment (pmhsfelix) [reply src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt]: We can add a `mapError` to the `Try` to do this mapping.

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt (hunk 11 above)
Comment (pmhsfelix) [reply src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt]: From the `withContext` docs:
```
Calls the specified suspending block with a given coroutine context, suspends until it completes, and returns the result.
```
So, it seems we are just changing the thread that gets blocked. Namely, the iteration will not proceed while the `getTextFromPage` doesn't complete.

File: src/main/kotlin/org/ionproject/integration/extractor/interface/PdfExtractor.kt (hunk 5 above)
Comment (pmhsfelix) [reply src/main/kotlin/org/ionproject/integration/extractor/interface/PdfExtractor.kt]: From the following sentence present at [https://github.com/tabulapdf/tabula-java/], I got the idea that it is possible to use Tabula as a library
```
You can also integrate tabula-java with any JVM language. For Java examples, see the tests folder.
```

File: src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt (hunk 13 above)
Comment (Ts1n) [reply src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt]: The tool is named [`IText`](https://github.com/itext/itext7)

File: src/main/kotlin/org/ionproject/integration/extractor/interface/PdfExtractor.kt (hunk 5 above)
Comment (Ts1n) [reply src/main/kotlin/org/ionproject/integration/extractor/interface/PdfExtractor.kt]: From what I was able to gather we should use `CommandLineApp` and that receives a file path. The alternative would be to use [PDFBox](https://pdfbox.apache.org/) that [tabula uses](https://github.com/tabulapdf/tabula-java/blob/master/pom.xml#L238), but that would negate the benefits of using `tabula`...

File: src/main/kotlin/org/ionproject/integration/extractor/implementations/ITextPdfExtractor.kt [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+import org.ionproject.integration.extractor.exceptions.PdfExtractorException
+import org.ionproject.integration.extractor.interfaces.PdfExtractor
+import org.ionproject.integration.utils.Try
+
+class ITextPdfExtractor : PdfExtractor {
+    /**
+     * Extract text data from pdf file locate at [pdfPath]
+     * @return CompositeException with list of exceptions in case of any error
+     */
+    override fun extract(pdfPath: String): Try<MutableList<String>> {
```
Comment (sscosta) [inline src/main/kotlin/org/ionproject/integration/extractor/implementations/ITextPdfExtractor.kt]: I think we should return Try<List< String>>

File: src/main/kotlin/org/ionproject/integration/extractor/implementations/ITextPdfExtractor.kt [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+import com.itextpdf.kernel.pdf.PdfReader
+import com.itextpdf.kernel.pdf.canvas.parser.PdfTextExtractor
+import org.ionproject.integration.extractor.exceptions.PdfExtractorException
+import org.ionproject.integration.extractor.interfaces.PdfExtractor
+import org.ionproject.integration.utils.Try
+
+class ITextPdfExtractor : PdfExtractor {
+    /**
+     * Extract text data from pdf file locate at [pdfPath]
+     * @return CompositeException with list of exceptions in case of any error
```
Comment (sscosta) [inline src/main/kotlin/org/ionproject/integration/extractor/implementations/ITextPdfExtractor.kt]: It also returns PDFExtractorException when path is empty

File: src/main/kotlin/org/ionproject/integration/extractor/implementations/ITextPdfExtractor.kt [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+                .map { reader -> PdfDocument(reader) }
+                .map { pdfDoc ->
+                    for (i in 1..pdfDoc.numberOfPages)
+                        data.add(PdfTextExtractor.getTextFromPage(pdfDoc.getPage(i)))
+                }
+                .map { data }
+                .mapError { PdfExtractorException("Itext cannot process file") }
+        } finally {
+            pdfReader
+                .match({ reader -> reader.close() }, {})
```
Comment (sscosta) [inline src/main/kotlin/org/ionproject/integration/extractor/implementations/ITextPdfExtractor.kt]: As we're not using the second function of match, i think it is equivalent to using the map function

File: build.gradle.kts [hunk 17]
```diff
@@ -24,6 +24,11 @@ dependencies {
     implementation("org.springframework.boot:spring-boot-starter-jdbc")
     implementation("org.jetbrains.kotlin:kotlin-reflect")
     implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
+    implementation("technology.tabula:tabula:1.0.3") {
+        exclude(group = "org.slf4j")
+    }
+    implementation("com.itextpdf:kernel:7.0.0")
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.3.5")
```
Comment (sscosta) [inline build.gradle.kts]: For now we will not need coroutines

**Review by sscosta — CHANGES_REQUESTED** [review]

**Review by sscosta — APPROVED** [review]

**Review by pmhsfelix — APPROVED** [review]
